### PR TITLE
Remove step and calls to process.nextTick

### PR DESCRIPTION
**Pull request by Algunenano**
- Removes unused dependency `sphericalmercator`.
- Removes usage of `step` and calls to `process.nextTick` that I think are no longer necessary.
- Updates generic-pool to 3.4.2.
- Some minor cleaning up with ES6 features.
- The changes in the tests are due to an error that was being hidden by step.

As I worked in the pool and step at the same time the commits cannot be picked individually.

### CODE REVIEW

File: lib/mapnik_backend.js [hunk 1]
```diff
@@ -22,7 +20,7 @@ function MapnikSource(uri, callback) {
     uri = this._normalizeURI(uri);
 
     if (uri.protocol && uri.protocol !== 'mapnik:') {
-        throw new Error('Only the mapnik protocol is supported');
+        callback(new Error('Only the mapnik protocol is supported'));
```
Comment (dgaubert) [inline lib/mapnik_backend.js]: Missing `return`

File: lib/mapnik_backend.js [hunk 2]
```diff
@@ -207,13 +190,11 @@ MapnikSource.prototype._close = function(callback) {
...
-        // NOTE: .drain() and .destoryAllNow() do
-        // not pass error args
-        pool.drain(function() {
-            pool.destroyAllNow(function() {
-                return callback();
-            });
+        const pool = this._pool;
+        const p = pool.drain()
+        .then(function() {
+            pool.clear();
```
Comment (dgaubert) [inline lib/mapnik_backend.js]: what's the difference between `.clear()` and `destroyAllNow()`??

File: lib/mapnik_backend.js [hunk 3]
```diff
@@ -226,52 +207,64 @@ MapnikSource.registerProtocols = function(tilelive) {
...
+    }
 
     // This is a string-based map file. Pass it on literally.
     if (this._uri.xml) return callback(null, this._uri.xml);
 
     // Load XML from file.
-    fs.readFile(path.resolve(source._uri.pathname), 'utf8', function(err, xml) {
-        if (err) return callback(err);
-        callback(null, xml);
+    fs.readFile(path.resolve(this._uri.pathname), 'utf8', (err, xml) => {
```
Comment (dgaubert) [inline lib/mapnik_backend.js]: if you don't want to do anything after reading the file, do better:
```
fs.readFile(path.resolve(this._uri.pathname), 'utf8', callback);
```
if you don't want to do anything but you want to wrap the original callback, for the sake of explicity:
```
fs.readFile(path.resolve(this._uri.pathname), 'utf8', function (err, xml) {
  if (err) {
    return callback(err);
  }
  return callback(null, xml);
});
```

File: lib/mapnik_backend.js [hunk 4]
```diff
@@ -226,52 +207,64 @@ MapnikSource.registerProtocols = function(tilelive) {
...
-            }
+MapnikSource.prototype._createPool = function(xml) {
+    if (this._pool) return;
+
+    const opts = {
+        max: this._uri.query.poolSize || process.env.UV_THREADPOOL_SIZE || require('os').cpus().length
+    };
+    const factory = {
+        create: () => {
+            return new Promise((resolve, reject) => {
```
Comment (dgaubert) [inline lib/mapnik_backend.js]: I understand your motivations to use promises here, but if you take a deeper look at what you are doing here:
```
. _populateInfo() [callback]-> .create() [promise] -> map.fromString() [callback]
```
The function `_populateInfo`, which uses callbacks, calls to `factory.create` which creates a promise and calls  `map.fromString` which is the original function that returns its value via callbacks. So we are using two ways to handle asynchronous code (callbacks and promises) but we are not taking advantage of promises at all, we are adding complexity for free. Obviously, the future (present to be honest) to handle asynchronous code are `promises`. But we need to transition to them in the right way. For that, we need to wrap or, better, reimplement the original function that produces asynchronous code to return promises. For instance:
```
. _populateInfo() [callback]-> .create() [callback] -> map.fromString() [promise]
```
then
```
. _populateInfo() [callback]-> .create() [promise] -> map.fromString() [promise]
```
and finally:
```
. _populateInfo() [promise]-> .create() [promise] -> map.fromString() [promise]
```
In conclusion to begin to use promises we must start from the bottom, that means, we need to reimplement `node-mapnik` to return promises (we can also wrap node mapnik with some [library](https://github.com/CartoDB/cartonik/blob/master/src/map-renderer.js#L10-L19) to move from callbacks to promises). Do you accept the challenge?? 😈 

For this PR, I encourage you to not use promises, please use callbacks instead. We are doing the same by removing `step` in `Windshaft-cartodb`, we are replacing `step` by callbacks. There is an exception to this rule, if you need to do perform the same asynchronous operation multiple times, as you did with this [PR](https://github.com/CartoDB/Windshaft/pull/632/files) to fetch column names for all layers.

File: lib/render.js [hunk 5]
```diff
@@ -48,36 +47,41 @@ function calculateMetatile(options) {
...
-                stats_tile = Object.assign(stats_tile, source_image.get_metrics());
-                tiles[key] = {
+    const tiles = {};
+    const err_num = 0;
+    let tile_num = 0;
+
+    meta.tiles.forEach(c => {
+        const key = [options.format, c[0], c[1], c[2]].join(',');
+        const encodeStartTime = Date.now();
+        const x = (c[1] - meta.x) * options.tileSize;
```
Comment (dgaubert) [inline lib/render.js:20]: ❤️

File: lib/render.js [hunk 6]
```diff
@@ -186,52 +190,42 @@ MapnikSource.prototype._renderMetatile = function(options, callback) {
     // Add reference to the source allowing debug/stat reporting to be compiled.
     options.source = source;
 
-    process.nextTick(function() {
-        // acquire can throw if pool is draining
```
Comment (dgaubert) [inline lib/render.js]: As you upgraded `generic-pool`, we need to review if acquire can throw errors. If not, please remove this ugly `try/catch`

File: lib/lockingcache.js [hunk 7]
```diff
@@ -65,21 +65,19 @@ LockingCache.prototype.clear = function() {
 
 LockingCache.prototype.trigger = function(id) {
     if (this.results[id] && this.results[id] !== true) {
-        process.nextTick(function() {
```
Comment (dgaubert) [inline lib/lockingcache.js:68]: ❤️

**Review by dgaubert — CHANGES_REQUESTED** [review]
This PR changes a lot of things, we need to test it intensively before proceeding.

Please, take a look at the comments.

File: lib/mapnik_backend.js (hunk 2 above)
Comment (Algunenano) [reply lib/mapnik_backend.js]: `destroyAllNow` was removed in the update. The [recommended way](https://github.com/coopernurse/node-pool#draining) now to drain when you don't want to wait for timeouts is to call clear.

File: lib/mapnik_backend.js (hunk 4 above)
Comment (Algunenano) [reply lib/mapnik_backend.js]: > In conclusion to begin to use promises we must start from the bottom, that means, we need to reimplement node-mapnik to return promises (we can also wrap node mapnik with some library to move from callbacks to promises). Do you accept the challenge?? smiling_imp

With the requirement of maintaining compatibility with both callback and promises, that's pretty tough to do in a reasonable amount of time in node-mapnik. Wanna create a project? I add promises and you join the modules between node-mapnik and Windshaft :smile:

> For this PR, I encourage you to not use promises, please use callbacks instead.

`generic-pool` removed the callback interface. So I can either revert the update, wrap the `acquire` and `release` with callbacks or leave the promises. I think either 1 or 3, but 2 is just going backwards. What do you think?

File: lib/render.js (hunk 6 above)
Comment (Algunenano) [reply lib/render.js]: It won't throw errors now, [just reject the promise](https://github.com/coopernurse/node-pool/blob/72b31a434c7b05ad879b2ace9830a9aa9fbba002/lib/Pool.js#L434).

But `map.render` can throw depending on the input. It's one of those functions that for input related errors (like invalid types in options) throws, but for other stuff (errors during the render), returns the error in the callback. So we still need the try-catch.

File: lib/mapnik_backend.js (hunk 4 above)
Comment (dgaubert) [reply lib/mapnik_backend.js]: > generic-pool removed the callback interface.

Whoa, what a pity. if `generic-pool` has other features that we can be interested in, go for 3. I not, go for 1.

Comment (Algunenano) [reply lib/mapnik_backend.js]: I've reverted to 2.5.4. Also I reworked _createMetatileCache and _renderMetatile to avoid the need to use `nextTick()`. Can you have another look, please?

File: lib/mapnik_backend.js [hunk 8]
```diff
@@ -226,57 +205,51 @@ MapnikSource.registerProtocols = function(tilelive) {
...
-    });
-    callback(null);
+        max: this._uri.query.poolSize || process.env.UV_THREADPOOL_SIZE || require('os').cpus().length
+    };
+
+    this._pool = Pool(factory);
 };
 
-MapnikSource.prototype._populateInfo = function(callback) {
-    var source = this;
```
Comment (dgaubert) [inline lib/mapnik_backend.js]: ❤️

File: lib/mapnik_backend.js [hunk 9]
```diff
@@ -175,29 +166,21 @@ MapnikSource.prototype._open = function(uri) {
...
-    }, function(err) {
-        if (err) return error(err);
-        source.open = true;
-        source.emit('open', null, source);
+    this._loadXML((err, xml) => {
+        if (err) return this.emit('open', err);
+
+        this._createMetatileCache(uri.query.metatileCache);
+        this._createPool(xml);
+        this._populateInfo((err_populate) => {
```
Comment (dgaubert) [inline lib/mapnik_backend.js]: Careful!!. You've just removed this callback. See https://github.com/CartoDB/tilelive-mapnik/pull/23/files#diff-92989f54c4804505d7ebdacce2a1a9b9R245

File: lib/mapnik_backend.js [hunk 10]
```diff
@@ -175,29 +166,21 @@ MapnikSource.prototype._open = function(uri) {
...
+            if (err_populate) return this.emit('open', err_populate);
+            this.open = true;
+            this.emit('open', null, this);
```
Comment (dgaubert) [inline lib/mapnik_backend.js]: Events don't work in the same way that callback to propagate errors. When an error happens you should emit the error event.

```
this.emit('error', err);
this.emit('open', value);
```
instead of
```
this.emit('open', err); // when error
this.emit('open', null, value); 
```

Comment (dgaubert) [inline lib/mapnik_backend.js]: Where are you handling `.on('open', function () {...}`

**Review by dgaubert — CHANGES_REQUESTED** [review]

File: lib/mapnik_backend.js (hunk 10 above)
Comment (Algunenano) [reply lib/mapnik_backend.js]: https://github.com/Algunenano/tilelive-mapnik/blob/1458acbcbb9e8028adf1d030792f58f8a0ce48b2/lib/mapnik_backend.js#L33 and https://github.com/Algunenano/tilelive-mapnik/blob/1458acbcbb9e8028adf1d030792f58f8a0ce48b2/lib/mapnik_backend.js#L45

Comment (Algunenano) [reply lib/mapnik_backend.js]: Yeah, but it's setup so all parameters passed to `emit` are relied as-is to a callback function, so in this case they work like that. See https://github.com/Algunenano/tilelive-mapnik/blob/1458acbcbb9e8028adf1d030792f58f8a0ce48b2/lib/mapnik_backend.js#L33

We could change it to use 2 different events, but that's how it was set up previously (I didn't change anything there).

Comment (dgaubert) [reply lib/mapnik_backend.js]: Oh my goodness! This module is full of surprises.

_circulen_ 🤷‍♂️

Comment (dgaubert) [reply lib/mapnik_backend.js]: Thanks!

File: lib/mapnik_backend.js [hunk 11]
```diff
@@ -175,29 +166,17 @@ MapnikSource.prototype._open = function(uri) {
...
-    }, function(err) {
-        if (err) return error(err);
-        source.open = true;
-        source.emit('open', null, source);
+    this._loadXML((err, xml) => {
+        if (err) return this.emit('open', err);
+
+        this._createMetatileCache(uri.query.metatileCache);
+        this._createPool(xml);
+        this._populateInfo();
```
Comment (dgaubert) [inline lib/mapnik_backend.js]: Just curious, why do you move `this.open = true` and `this.emit('open', null, this)` to `._populateInfo()`?

File: lib/mapnik_backend.js [hunk 12]
```diff
@@ -226,57 +201,50 @@ MapnikSource.registerProtocols = function(tilelive) {
...
+
     // This is a string-based map file. Pass it on literally.
     if (this._uri.xml) return callback(null, this._uri.xml);
 
     // Load XML from file.
-    fs.readFile(path.resolve(source._uri.pathname), 'utf8', function(err, xml) {
-        if (err) return callback(err);
-        callback(null, xml);
-    });
+    fs.readFile(path.resolve(this._uri.pathname), 'utf8', callback);
```
Comment (dgaubert) [inline lib/mapnik_backend.js:211]: That's cool, but we never reach this code 👿. We always use `this._uri.xml`.

Comment (dgaubert) [reply lib/mapnik_backend.js:211]: FYI

File: lib/render.js [hunk 13]
```diff
@@ -186,56 +191,52 @@ MapnikSource.prototype._renderMetatile = function(options, callback) {
...
+            options.y = meta.y;
+            options.variables.zoom = options.z;
+            map.resize(meta.width, meta.height);
+            map.extent = meta.bbox;
+            try {
+                source._stats.render++;
+                const renderStats = {};
+                const renderStartTime = Date.now();
+                map.render(image, options, (err, image) => {
+                    this._pool.release(map);
```
Comment (dgaubert) [inline lib/render.js]: Regarding the comment: `Release after the .render() callback returned	to avoid mapnik errors.` to justify the `.nectTick()`. Why do we avoid mapnik errors now?

File: lib/render.js [hunk 14]
```diff
@@ -186,56 +191,52 @@ MapnikSource.prototype._renderMetatile = function(options, callback) {
...
+    } catch (err) {
+        return callback(err, keys);
+    }
+
     // Return a list of all the tile coordinates that are being rendered
     // as part of this metatile.
-    return meta.tiles.map(function(tile) {
-        return options.format + ',' + tile.join(',');
-    });
+    return keys;
```
Comment (dgaubert) [inline lib/render.js]: Is that still needed?

File: package.json [hunk 15]
```diff
@@ -27,11 +27,9 @@
         "Blake Thompson <flippmoke>"
     ],
     "dependencies": {
-        "step": "~0.0.5",
```
Comment (dgaubert) [inline package.json:30]: :heart:

File: lib/render.js (hunk 14 above)
Comment (Algunenano) [reply lib/render.js]: Yes, it's needed because the function passed to `LockingCache` has to return the ids of the tiles generated.

File: lib/render.js (hunk 13 above)
Comment (Algunenano) [reply lib/render.js]: I don't know why that was there; to my knowledge there is nothing in node-mapnik that requires the map object to live for an extra tick.
`_pool.release(map)` just does `delete map` (plus whatever it needs to do internally in the pool) and, since you are no longer using `map`, it should be perfectly safe. If it wasn't that would mean there is some memory being incorrectly shared by `map` and `err` or `image`. That case would be a (big) bug in node-mapnik and it should be fixed there instead of artificially delay the resource release.

File: lib/mapnik_backend.js (hunk 12 above)
Comment (Algunenano) [reply lib/mapnik_backend.js:211]: I know, but just in case we decide to change it in the future.

File: lib/render.js (hunk 14 above)
Comment (dgaubert) [reply lib/render.js]: Ok, I didn't notice the `return` when calling `_renderMetatile()`. I think, we can extract:
```
const keys = meta.tiles.map(function(tile) {
   return options.format + ',' + tile.join(',');
});
```
to a method to be called from `_renderMetatile` and `LockingCache`'s generator.  Don't you mind give a try?

File: lib/render.js (hunk 6 above)
Comment (dgaubert) [reply lib/render.js]: IMHO, this one of biggest pains of `node-mapnik`, it's an anti-pattern in Node.js ecosystem. When we have the chance, I'd like to change that behavior: any asynchronous function MUST propagate `errors` via `callback` OR rejecting the `promise`.

File: lib/render.js (hunk 13 above)
Comment (dgaubert) [reply lib/render.js]: Thanks for the explanation.

File: lib/mapnik_backend.js (hunk 11 above)
Comment (Algunenano) [reply lib/mapnik_backend.js]: In insight it was probably an error since it was clearer having it in _open. I'm getting kind of tired of mixing events for callback so I'll try to remove them and see how it gets.

Comment (dgaubert) [reply lib/mapnik_backend.js]: Looking forward to seeing how it looks like.

File: lib/render.js (hunk 14 above)
Comment (Algunenano) [reply lib/render.js]: I tried but since stopped because to calculate the keys you need the output from `calculateMetatile` and that was being called in `_renderMetatile`. I'll see if I can move that call into createMetaTile instead so it doesn't need to get called twice.

File: lib/mapnik_backend.js (hunk 11 above)
Comment (Algunenano) [reply lib/mapnik_backend.js]: Done in 561800e.

File: lib/render.js (hunk 14 above)
Comment (Algunenano) [reply lib/render.js]: Done in bb20d50.

File: lib/mapnik_backend.js [hunk 16]
```diff
@@ -290,21 +260,65 @@ MapnikSource.prototype._populateInfo = function(callback) {
...
-        source._info = info;
-        source._pool.release(map);
-        callback(null)
+        this._info = info;
+        this._pool.release(map);
+        return callback(null);
     });
 };
 
+exports['calculateMetatile'] = calculateMetatile;
```
Comment (dgaubert) [inline lib/mapnik_backend.js:267]: Once you moved this function to `mapnik_nackend`, do we need to export it? I guess for testing porpuses, doesn't it?

**Review by dgaubert — APPROVED** [review]
Lots of changes. Before merging, I'd be nice to link this branch to `windshaft` and then to `windshaft-cartodb` and perform some test and benchmark if possible.

Really nice work. I appreciate much the effort that you put here. This module is really complex and any change here could lead to a performance degradation.

File: lib/mapnik_backend.js (hunk 16 above)
Comment (Algunenano) [reply lib/mapnik_backend.js:267]: Yes, just for testing.

**Comment by Algunenano** [comment]
> Lots of changes. Before merging, I'd be nice to link this branch to windshaft and then to windshaft-cartodb and perform some test and benchmark if possible.

I have other changes (node-mapnik update) that I want to test too in Windshaft. Do you mind if I merge this and release it so I can test everything in Windshaft? Keep in mind that Windshaft pins the tilelive-mapnik release (no `~` or `^`) so making an extra release (or 2 if we need to undo some changes) won't propagate into production unless we want to.

**Comment by dgaubert** [comment]
Proceed. Make extra-release if needed.
